### ISSUE TITLE
Clean up continuous state code

### ISSUE THREAD

**Issue by jstac** [jstac/sdfs_via_autodiff#5]
Hi @JunnanZ , I notice some repeated code in https://github.com/jstac/sdfs_via_autodiff/tree/main/code/ssy/continuous_junnan

Can this be unified?

@HumphreyYang and @Smit-create are starting to work on the code base.   One of their first tasks will be to make sure that the SSY continuous state code is optimized.  For example, they are looking at different interpolation methods in JAX. 

It would be great if you could assist by eliminating repeated code, documenting and trying to make the approximation / interpolation part "plug-and-play", so that we can test which is best.

**Comment by JunnanZ** [comment]
> Can this be unified? 
> It would be great if you could assist by eliminating repeated code, documenting and trying to make the approximation / interpolation part "plug-and-play", so that we can test which is best.

Yeah, I've been meaning to do that. I will clean it up in a day or two.

**Comment by JunnanZ** [comment]
Hi @HumphreyYang and @Smit-create, I have removed most of the repeated code and added more documentation in d497e7c. Please let me know if any part is confusing. 

The interpolation method (`lin_interp` from `utils.py`) only appears once in each inner function (`Kg_vmap_mc` and `Kg_vmap_quad`). It's not exactly plug-and-play but close. As a first step, you may want to try different interpolation methods and compare them with `lin_interp`.

To be honest, the code is not very elegant, but I don't have any good idea that can lead to a major improvement. If you have any good ideas, feel free to open PRs.

Finally, in order for the computation to fit in memory, I combine `jax.vmap` with `jax.lax.map` but performance suffers because of this. In theory, `vmap` combined with `jax.jit` can automatically optimize memory use and avoid generating large intermediate arrays (see https://github.com/google/jax/discussions/9505). However, I could never make this work for those two `Kg_vmap` functions. Please let me know if you have any idea how to solve this and avoid `jax.lax.map`.

**Comment by Smit-create** [comment]
Perfect, thanks @JunnanZ. I have tested the code for `lin_interp` and it works perfect for some random tests that I tried using:
```py
for i in range(20):
    np.random.seed(i*10+100)
    points = np.random.random((10000, 2))
    linear_interp = eval_linear(grid, values, points) # 10000 vector
    out_3 = lin_interp(points.T, values, (np.linspace(*grid[0]), np.linspace(*grid[1])))
    assert np.max(np.abs(out_3 - linear_interp)) < 1e-12
```

**Comment by Smit-create** [comment]
I tried the same for 3 and 4 dimensions and everything works fine for me.

<details>
<summary> 3-dim test</summary>
<p>

```py
def f(x, y, w): 
    p = np.sin(np.sqrt(x**2+y**2+0.00001))/np.sqrt(x**2+y**2+0.00001)
    p += np.sin(w**2 +0.00001)/(w**2+ 0.00001)
    return p

# uniform cartesian grid
grid = UCGrid((-1.0, 1.0, 10), (-4.0, 1.0, 10), (-1.0, 3.0, 10))

# get grid points
gp = nodes(grid)

v = np.empty((10,10,10))
for i in range(10):
    for j in range(10):
        for k in range(10):
                v[i,j,k]=f(gp[i,0], gp[j,1], gp[k,2])

for i in range(20):
    np.random.seed(i*10+100)
    dim = 3
    points = np.random.random((10000, dim))
    linear_interp = eval_linear(grid, v, points)
    out_3 = lin_interp(points.T, v, (np.linspace(*grid[0]), np.linspace(*grid[1]),
                                    np.linspace(*grid[2])), dim)
    assert np.max(np.abs(out_3 - linear_interp)) < 1e-12
```
</p>
</details>

<details>
<summary> 4-dim test</summary>
<p>

```py
def f(x, y, w,z): 
    p = np.sin(np.sqrt(x**2+y**2+0.00001))/np.sqrt(x**2+y**2+0.00001)
    p += np.sin(w**2 + z +0.00001)/(w**2+ z + 0.00001)
    return p

# uniform cartesian grid
grid = UCGrid((-1.0, 1.0, 10), (-4.0, 1.0, 10), (10.0, 12.0, 10), (-1.0, 3.0, 10))

# get grid points
gp = nodes(grid)

v = np.empty((10,10,10,10))
for i in range(10):
    for j in range(10):
        for k in range(10):
            for l in range(10):
                v[i,j,k,l]=f(gp[i,0], gp[j,1], gp[k,2], gp[l,3])

for i in range(20):
    np.random.seed(i*10+100)
    dim = 4
    points = np.random.random((10000, dim))
    linear_interp = eval_linear(grid, v, points)
    out_3 = lin_interp(points.T, v, (np.linspace(*grid[0]), np.linspace(*grid[1]),
                                    np.linspace(*grid[2]), np.linspace(*grid[3])), dim)
    assert np.max(np.abs(out_3 - linear_interp)) < 1e-12
```
</p>
</details>

**Comment by JunnanZ** [comment]
Fixed in 429a2a6

I moved the code for generating `x_3d` inside the jit'd function $T$ and this significantly improved compilation time for Newton't method. I have no idea why.

In 429a2a6 I also added more tests when comparing two function factories.

Anyway, this issue can be closed. We can discuss performance related ideas in #10.
